sklearn/tree/_splitter.pyx

```diff
@@ -816,20 +945,46 @@ cdef class DensePartitioner:
         SIZE_t best_pos,
         double best_threshold,
         SIZE_t best_feature,
+        SIZE_t best_n_missing,
     ) noexcept nogil:
         """Partition samples for X at the best_threshold and best_feature."""
         cdef:
-            SIZE_t p = self.start
-            SIZE_t partition_end = self.end
+            SIZE_t start = self.start
+            SIZE_t end = self.end
+            SIZE_t p, n_missing, partition_end
             SIZE_t[::1] samples = self.samples
             const DTYPE_t[:, :] X = self.X
-
-        while p < partition_end:
-            if X[samples[p], best_feature] <= best_threshold:
+            SIZE_t end_non_missing = self.end - best_n_missing
+
+        # Move missing values to the end
+        if best_n_missing != 0:
+            n_missing = 0
+            p, partition_end = start, end - 1
```

Comment:
Comparing with the previous implementation: is it possible that we miss a value here?

```suggestion
            p, partition_end = start, end
```

Reply:
I redid the logic in [`d12f336` (#23595)](https://github.com/scikit-learn/scikit-learn/pull/23595/commits/d12f336f7b55e962e72150eae1483758f3453563) so it looks more like the non-missing case. It should perform better because it only needs to go through the data once.